src/Sylius/Behat/Client/RequestBuilder.php

```diff
@@ -0,0 +1,77 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Behat\Client;
+
+use Symfony\Component\HttpFoundation\File\UploadedFile;
+
+final class RequestBuilder
+{
+    private string $uri;
+    private string $method;
+    private array $parameters = [];
+    private array $headers = [];
+    private array $content = [];
+    private array $files = [];
+
+    private function __construct(string $uri, string $method)
+    {
+        $this->uri = $uri;
+        $this->method = $method;
+    }
+
+    public static function create(string $uri, string $method): self
+    {
+        return new self($uri, $method);
+    }
+
+    public function withHeader(string $name, string $value): self
+    {
+        $this->headers[$name] = $value;
+
+        return $this;
+    }
```

Comment:
I'm wondering if there would be any benefit of making these calls immutable
```suggestion
    public function withHeader(string $name, string $value): self
    {
        $builder = self::create($this->uri, $this->method);
        $builder->headers[$name] = $value;

        return $builder;
    }
```

Reply:
I'm thinking that the builder is kind of mutable pattern (but I might be wrong). Your example implementation will fail with the use case:
```php
$builder
  ->withHeader('header1', 'value1')
  ->withHeader('header2', 'value2')
```

We could change builder implementation like this: (postponed most of the methods)

```php
<?php

final class RequestBuilder
{
    private function __construct(
        private string $uri,
        private string $method,
        private array $parameters = [],
        private array $headers = [],
        private array $content = [],
        private array $files = [],
    )
    {
    }

    public static function create(string $uri, string $method): self
    {
        return new self($uri, $method);
    }

    public function withHeader(string $name, string $value): self
    {
        $builder = self::create($this->uri, $this->method, $this->parameters, $this->headers /* ... */);
        $builder->headers[$name] = $value;
        
        return $builder;
    }

    public function build(): RequestInterface
    {
        return new Request(
            $this->uri,
            $this->method,
            $this->parameters,
            $this->headers,
            $this->content,
            $this->files,
        );
    }
}
```

But i don't know if this isn't a little bit of overenginnering